### PR TITLE
[datetime] fix(DateInput): make popover focusable via keyboard

### PR DESCRIPTION
**Pull request by michael-yx-wu**
#### Changes proposed in this pull request:

Pressing Tab when focused on a `DateInput`'s text field will focus the popover content. Previously, pressing tab would focus the next keyboard-focusable element on the page. Since the popover is rendered with a portal by default, this would usually result in the popover closing.

Does _not_ autofocus the popover to continue to allow users to manually type in a date string if they prefer not to use the popover.

#### Screenshot

Before:

After:

Worth noting that the element tab order in the DatePicker seems weird
1. Left/right arrows
2. Month dropdown
3. Year dropdown

I would expect this order instead:
1. Left arrow
2. Month dropdown
3. Year dropdown
4. Right arrow

### CODE REVIEW

File: packages/docs-app/src/examples/core-examples/popoverDismissExample.tsx [hunk 1]
```diff
@@ -36,6 +36,8 @@ export class PopoverDismissExample extends React.PureComponent<
         return (
             <Example options={false} {...this.props}>
                 <Popover
+                    autoFocus={false}
+                    enforceFocus={false}
```
Comment (michael-yx-wu) [inline packages/docs-app/src/examples/core-examples/popoverDismissExample.tsx:40]: This Popover example continuously reopens by design, but now that the `enforceFocus` behavior is fixed, it will reset the scroll position to this part of the page, making it difficult to interact with the other examples. Happy to do in a separate PR if you prefer

Comment (adidahiya) [reply packages/docs-app/src/examples/core-examples/popoverDismissExample.tsx:40]: yeah I was mostly just punting this problem because it's fixed in the Popover2 example. no big deal, we can make the change here

**Review by adidahiya — COMMENTED** [review]
how do you get back to the input after tabbing into the popover? shift+tab doesn't seem to work:

**Comment by michael-yx-wu** [comment]
Shift+tab interactions should be fixed now:

File: packages/datetime/src/dateInput.tsx [hunk 2]
```diff
@@ -421,6 +425,7 @@ export class DateInput extends AbstractPureComponent2<DateInputProps, IDateInput
             this.handleDateChange(nextDate, true, true);
         } else if (e.which === Keys.TAB && this.state.isOpen) {
             this.getFirstTabbableElement()?.focus();
+            e.preventDefault();
```
Comment (michael-yx-wu) [inline packages/datetime/src/dateInput.tsx:411]: Necessary to prevent focusing the 2nd focusable element

Comment (adidahiya) [reply packages/datetime/src/dateInput.tsx:411]: sounds like this should be a code comment

```suggestion
            // necessary to prevent focusing the 2nd focusable element
            e.preventDefault();
```

File: packages/datetime/src/dateInput.tsx [hunk 3]
```diff
@@ -419,15 +423,21 @@ export class DateInput extends AbstractPureComponent2<DateInputProps, IDateInput
         if (e.which === Keys.ENTER) {
             const nextDate = this.parseDate(this.state.valueString);
             this.handleDateChange(nextDate, true, true);
-        } else if (e.which === Keys.TAB) {
-            this.setState({ isOpen: false });
+        } else if (e.which === Keys.TAB && e.shiftKey) {
+            this.handleClosePopover();
```
Comment (adidahiya) [inline packages/datetime/src/dateInput.tsx:407]:
```suggestion
            // close popover on SHIFT+TAB key press
            this.handleClosePopover();
```

File: packages/datetime/src/dateInput.tsx [hunk 4]
```diff
@@ -463,24 +473,33 @@ export class DateInput extends AbstractPureComponent2<DateInputProps, IDateInput
...
+                    this.inputElement?.focus();
+                    e.preventDefault();
+                }
                 this.handleClosePopover();
             }
         } else if (relatedTarget != null) {
-            this.unregisterPopoverBlurHandler();
+            this.unregisterPopoverBlurHandlers();
+            this.firstTabbableElement = this.getFirstTabbableElement();
+            this.firstTabbableElement?.addEventListener("blur", this.handlePopoverBlur);
```
Comment (adidahiya) [inline packages/datetime/src/dateInput.tsx]: this doesn't seem exactly right... I feel like the popover should stay open after a SHIFT+TAB keypress on the first tabbable element?

for example, CLICK, TAB, SHIFT+TAB interaction here:

**Review by adidahiya — CHANGES_REQUESTED** [review]
almost there

File: packages/datetime/src/dateInput.tsx (hunk 4 above)
Comment (adidahiya) [reply packages/datetime/src/dateInput.tsx]: I could go either way, though... open to other ideas

File: packages/datetime/src/dateInput.tsx (hunk 2 above)
Comment (michael-yx-wu) [reply packages/datetime/src/dateInput.tsx:411]: ~~I moved some things around and don't need to this anymore~~

just kidding, I still need this. I'll add the comment

File: packages/datetime/src/dateInput.tsx [hunk 5]
```diff
@@ -419,70 +424,55 @@ export class DateInput extends AbstractPureComponent2<DateInputProps, IDateInput
...
-        // a blur handler, so that when:
-        // * user tabs to the next element, popover closes
-        // * focus moves to element within popover, popover stays open
-        const lastTabbableElement = numOfElements > 0 ? tabbableElements[numOfElements - 1] : null;
-
-        return lastTabbableElement as HTMLElement | null;
+    private getKeyboardFocusableElements = (): HTMLElement[] => {
+        const elements: HTMLElement[] = Array.from(
+            this.popoverContentElement?.querySelectorAll(
+                "button:not([disabled]),input,[tabindex]:not([tabindex='-1'])",
```
Comment (michael-yx-wu) [inline packages/datetime/src/dateInput.tsx:422]: This is a subset of the selector used in `Overlay`. I wonder if it's worth sharing that in the future

**Comment by michael-yx-wu** [comment]
Makes sense. I'm now keeping the popover open when switching focus to the input. Also fixed the issue of tabbing in the popover focusing the browser address bar.

**Comment by michael-yx-wu** [comment]
Will fix tests

File: packages/datetime/test/dateInputTests.tsx [hunk 6]
```diff
@@ -92,7 +92,11 @@ describe("<DateInput>", () => {
         assert.isFalse(wrapper.find(Popover).prop("isOpen"));
     });
 
-    it("Popover closes when tabbing on first day of the month", () => {
+    // Skipping because simulate just invokes the function passed to React's "on<EventName>" prop
+    // and doesn't actually simulate anything. Properly testing would require running with an actual
+    // browser and focusing specific elements via the DOM API. This would require changing the Karma
+    // config to run with Chrome instead of ChromeHeadless.
+    it.skip("Popover closes when tabbing on first day of the month", () => {
```
Comment (michael-yx-wu) [inline packages/datetime/test/dateInputTests.tsx:99]: I do have a working local version but could only get it to work by switching the Karma browser from `ChromeHeadless` to `Chrome`. The problem with simulate is that there's no way to get the `focusin` event handler functions to run because React doesn't provide an `onFocusIn` prop, only `onFocus`. If I use the DOM API instead and call `Element#focus()`, the browser will fire a `focusin` event on my behalf. I'm not sure if that's an anti-pattern for the unit tests.

Comment (adidahiya) [reply packages/datetime/test/dateInputTests.tsx:99]: ok, this is fine to skip for now. testing focus behavior is tricky in these test suites

File: packages/datetime/src/dateInput.tsx [hunk 7]
```diff
@@ -301,6 +305,8 @@ export class DateInput extends AbstractPureComponent2<DateInputProps, IDateInput
         const { popoverProps = {} } = this.props;
         popoverProps.onClose?.(e);
         this.setState({ isOpen: false });
+        this.startFocusBoundaryElement?.removeEventListener("focusin", this.handleStartFocusBoundaryFocusIn);
```
Comment (adidahiya) [inline packages/datetime/src/dateInput.tsx]: why do we need to use this ref approach and manually bind focusin event handlers?  why can’t it be declarative like normal React code `<div onFocusIn={this.handleEndFocusBoundaryFocusIn} />` ?

Comment (michael-yx-wu) [reply packages/datetime/src/dateInput.tsx]: React doesn't support `onFocusIn`, but it looks like their `onFocus` prop listens to the `focusin` event under the hood (https://github.com/facebook/react/issues/6410#issuecomment-671499208). I switched to this approach and everything still works as expected.

**Review by adidahiya — APPROVED** [review]
